src/io/xmap.rs

```diff
@@ -0,0 +1,1357 @@
+            Ok(false) => None,
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+/// A container with CMAP header and records stored in B-trees.
+///
+/// Each tree corresponds to one reference contig and is ordered by start position.
+#[derive(Debug, new, PartialEq)]
+pub struct Container {
+    header: Vec<String>,
+    pos_trees: HashMap<u32, BTreeMap<u64, Record>>,
+    //     pos_trees: HashMap<u32, BTreeMap<u32, Rc<Record>>>,
+}
+
+impl Container {
+    /// Constructs a new `Container` from path.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let reader = Reader::from_path(path)?;
+        reader.into_container()
+    }
+
+    /// Returns the subset of records contained in given region in a contig.
+    pub fn subset_filter_by_region(
+        &mut self,
+        region_contig: u32,
+        region_start: u64,
+        region_end: u64,
+    ) -> Result<Self> {
+        let hits = match self.pos_trees.get_mut(&region_contig) {
+            Some(hit_tree) => hit_tree.range(region_start..region_end),
+            None => Default::default(),
+        };
+        let mut tree = BTreeMap::new();
+        for entry in hits {
+            tree.insert(*entry.0, entry.1.clone());
+        }
+        Ok(Container {
+            header: self.header().clone(),
+            pos_trees: HashMap::from([(region_contig, tree)]),
+        })
+    }
+
+    /// Returns the subset of records contained in given contig.
+    pub fn subset_filter_by_contig(&mut self, region_contig: u32) -> Result<Self> {
+        let hits = match self.pos_trees.get_mut(&region_contig) {
+            Some(hit_tree) => hit_tree.clone(),
+            None => BTreeMap::default(),
+        };
+        Ok(Container {
+            header: self.header().clone(),
+            pos_trees: HashMap::from([(region_contig, hits)]),
+        })
+    }
+
+    /// Returns the range over records contained in a given region in a contig.
+    pub fn range_filter_by_region(
+        &mut self,
+        region_contig: u32,
+        region_start: u64,
+        region_end: u64,
+    ) -> Result<Range<u64, Record>> {
+        let hits = match self.pos_trees.get_mut(&region_contig) {
+            Some(hit_tree) => hit_tree.range(region_start..region_end),
+            None => Default::default(),
+        };
+        Ok(hits)
+    }
+
+    /// Returns the range over records contained in given contig.
+    pub fn range_filter_by_contig(&mut self, region_contig: u32) -> Result<Range<u64, Record>> {
+        let hits = match self.pos_trees.get_mut(&region_contig) {
+            Some(hit_tree) => hit_tree.range(1..),
+            None => Default::default(),
+        };
+        Ok(hits)
+    }
+
+    /// Returns header.
+    pub fn header(&self) -> &Vec<String> {
+        &self.header
+    }
+
+    /// Returns B-trees of start positions (for each contig).
+    pub fn trees(&mut self) -> HashMap<u32, BTreeMap<u64, Record>> {
+        self.pos_trees.clone()
+    }
```

Comment:
Not needed, why would one want a clone of this?

Reply:
I removed it.